enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,617 @@
+```
+#### Variation [optional]
+
+### API Extensions
+
+A) Set(Replace)/Append/Delete HTTP headers in the global section of haproxy.config via IngressController's Spec.
+The existing Ingress Controller API `HTTPHeaders *IngressControllerHTTPHeaders` is extended by adding an field for action to be performed,
+and field for headers to be set/replaced/appended/deleted.
+
+```go
+// IngressControllerSpec is the specification of the desired behavior of the
+// IngressController.
+type IngressControllerSpec struct {
+    // httpHeaders defines policy for HTTP headers.
+    //
+    // If this field is empty, the default values are used.
+    //
+    // +optional
+    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
+}
+```
+
+```go
+
+// IngressControllerHeadersManipulation specifies options for set/replace/append/delete headers.
+type IngressControllerHeadersManipulation struct {
+    Response []IngressControllerSetReplaceAppendDeleteHTTPHeaders `json:"response,omitempty"`
+}
+
+```
+
+```go
+// IngressControllerHeadersManipulation specifies options for set/replace/append/delete headers.
+// +union
+type IngressControllerSetReplaceAppendDeleteHTTPHeaders struct {
+    // action lets you specify headers for Set or Delete.
+    // +unionDiscriminator
+    // +kubebuilder:validation:Enum:=Set;Delete 
+	//+kubebuilder:validation:Required
+    Action IngressControllerSetHTTPHeaderAction `json:"action,omitempty"`
+
+    // name specifies a header name.  Its value must be a valid HTTP header
+    // name as defined in RFC 2616 section 4.2.
+    // +kubebuilder:validation:Optional
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=256
+    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+    // +optional
+    Name *string `json:"name"`
+
+    // Set defines HTTP headers that should be set or replaced or appended.
+    // If this field is empty, no headers are added.
+    //
+    // Note that this option only applies to cleartext HTTP connections
+    // and to secure HTTP connections for which the ingress controller
+    // terminates encryption (that is, edge-terminated or reencrypt
+    // connections).  Headers cannot be set(replace)/append for TLS passthrough
+    // connections.
+    //
+    // +kubebuilder:validation:Optional
+    // +optional
+    Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`
+
+    // delete defines HTTP headers that should be deleted.
+    // If this field is empty, no headers are deleted.
+    //
+    // Note that this option only applies to cleartext HTTP connections
+    // and to secure HTTP connections for which the ingress controller
+    // terminates encryption (that is, edge-terminated or reencrypt
+    // connections).  Headers cannot be set(replace)/append for TLS passthrough
+    // connections.
+    //
+    // +kubebuilder:validation:Optional
+    // +optional
+    Delete *IngressControllerDeleteHTTPHeader `json:"delete,omitempty"`
```

Comment:
Why `Delete` union member is needed? It doesn't have any data to carry.

Reply:
It can be omitted but IMO it is clearer to include it.

Reply:
api changes are updated